### ISSUE TITLE
Allow file actions priorization

### ISSUE THREAD

**Issue by nickvergessen** [nextcloud/server#5065]
### Steps to reproduce
1. Enable richdocuments to have rich editing for odt files
2. Enable files_texteditor because you want txt files to be simple
3. Open an txt file: richdocuments starts

I think it would be nice to have something like a priority of the actions. Currently the last one wins, and since richdocuments is after files_texteditor in the alphabet, files_texteditor has no way to ever win this.

Proposal is to add something like priority to: https://github.com/nextcloud/server/blob/2b794902014264a3793daf42a2ecfa95a1a07342/apps/files/js/fileactions.js#L144-L144

and then:
```js
			if (!!this.actions[mime][name] && this.actions[mime][name].priority < actionSpec.priority) {
				return;
			}
```

This would allow the files_texteditor to register for text files, while richdocuments can also, so in case files_texteditor is not there, you can still edit them.

Alternate approach would be hacking into richdocuments to only register, when files_texteditor is not there.

cc @MorrisJobke @LukasReschke 

cc @RealKelsar for reporting it in IRC

**Comment by MorrisJobke** [comment]
> This would allow the files_texteditor to register for text files, while richdocuments can also, so in case files_texteditor is not there, you can still edit them.

I like that approach more. The cleaner hack. But still feels like a hack.

**Comment by LukasReschke** [comment]
I'd rather prefer a "Open with" dialogue but yay, JavaScript 😉

**Comment by szaimen** [comment]
I suppose this issue is still valid? If not, please close this issue!

**Comment by susnux** [comment]
There are priorities for file actions now, but we still have two problems:

1. Actions that register as default actions -> if there are multiple we use the most important one without letting the user choose.
2. multiple handlers on the viewer (e.g. richdocuments and text both use the viewer)

1 is a server issue (this on) and 2 is a viewer issue and handled here: https://github.com/nextcloud/viewer/issues/2393

**Comment by Dante4** [comment]
Got here from https://github.com/nextcloud/server/issues/36801

Is there any news regarding this? 
In version 27 left click on file was opening detail menu, which is what I thought is default action, but current version instead always download file, which is slightly annoying change because I mostly use Nextcloud for web sharing, not to download this 50gb file-thank-you-very-much-need-to-cancel-download-agein

**Comment by Jaakurin** [comment]
@Dante4 I second this. The ability to change the default single-click action to file/folder selection would save a great deal of network bandwidth for many.

**Comment by ne0YT** [comment]
this really just needs to be customizable!

**Comment by akshansh1998** [comment]
yeah, I second this. I am so frustruted by keep downloading files, when just intend to open them or sometimes, just by accidently single clicking on files, it downloads them

**Comment by kmish85** [comment]
hello, i've made temporary unsupported workaround that will work with standalone install and breaks after each update, so keep in mind, to restore it after each update (or try to implement in in the source).
solution tested on nextcloud 32.0.0 (probably will work with other versions around)
so, we have to modify 
/var/www/nextcloud/dist/files-main.js
1. beautify the file;
2. somewhere in the beginning add a global variable
let lastTap = 0;
3. somewhere around line 3806 find the function (method) line:
execDefaultAction(e) {
4. in the end of the body add/modify after e.stopPropagation()
e.stopPropagation();	//replace comma to semicolon
/* kmish begins */
const currentTime = new Date().getTime();
const tapLength = currentTime - lastTap;
// Double-click detection
if (tapLength < 300 /*doubleClickDelay*/ && tapLength > 0)
	this.defaultFileAction.exec(this.source, this.currentView, this.currentDir)
else	
	Ge?.enabled?.([this.source], this.currentView) && Ge.exec(this.source, this.currentView, this.currentDir);
lastTap = currentTime;
/* kmish ends */
5. optionally minify js;
6. replace your file;
7. hardrefresh the browser to update js.

Hope it helps.

<img width="785" height="1203" alt="Image" src="https://github.com/user-attachments/assets/847ad6af-435b-411b-ad12-1d3bd2b4c2ea" />

**Comment by ne0YT** [comment]
> hello, i've made temporary unsupported workaround that will work with standalone install and breaks after each update, so keep in mind, to restore it after each update (or try to implement in in the source). solution tested on nextcloud 32.0.0 (probably will work with other versions around) so, we have to modify /var/www/nextcloud/dist/files-main.js
> 
> 1. beautify the file;
> 2. somewhere in the beginning add a global variable
>    let lastTap = 0;
> 3. somewhere around line 3806 find the function (method) line:
>    execDefaultAction(e) {
> 4. in the end of the body add/modify after e.stopPropagation()
>    e.stopPropagation();	//replace comma to semicolon
>    /* kmish begins _/
>    const currentTime = new Date().getTime();
>    const tapLength = currentTime - lastTap;
>    // Double-click detection
>    if (tapLength < 300 /_doubleClickDelay_/ && tapLength > 0)
>    this.defaultFileAction.exec(this.source, this.currentView, this.currentDir)
>    else
>    Ge?.enabled?.([this.source], this.currentView) && Ge.exec(this.source, this.currentView, this.currentDir);
>    lastTap = currentTime;
>    /_ kmish ends */
> 5. optionally minify js;
> 6. replace your file;
> 7. hardrefresh the browser to update js.
> 
> Hope it helps.
> 
> <img alt="Image" width="785" height="1203" src="https://private-user-images.githubusercontent.com/213068594/503253593-847ad6af-435b-411b-ad12-1d3bd2b4c2ea.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE2NTYyNzcsIm5iZiI6MTc2MTY1NTk3NywicGF0aCI6Ii8yMTMwNjg1OTQvNTAzMjUzNTkzLTg0N2FkNmFmLTQzNWItNDExYi1hZDEyLTFkM2JkMmI0YzJlYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAyOFQxMjUyNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNGE4MDlhM2I4Mjk0YmE2MTk3MDVlMjBhOGU2NjE4M2VjY2ZiMmM1MjI4NmE5YmIxNTQ4NjIwODRlYTgyM2NlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.VHibeag95R4mBdFzaUw-rV2Qm80jGCPdOJ7eKYj5phY">

would be great to have a script to temporarily include it in our update-script.

**Comment by mitchell-accipio** [comment]
Bumping this as *incredibly* frustrating if a user clicks a 500MiB ZIP file and it starts *immediately* downloading it. Lots of complaints on this. Opening the details pane should be the default. If NC doesn't have a handler for a file, selecting it doesnt mean I want to download it.

**Comment by Noctisik** [comment]
Still works pretty well on 34.0.1
```js
                 e.preventDefault(), 
								e.stopPropagation(); // replaced with semicolon 
								// this.defaultFileAction.exec({
								/* kmish  begins */

								e.preventDefault(),e.stopPropagation();

								const currentTime=Date.now();
								const tapLength=currentTime-this.lastTap;
								const actionContext={
									nodes:[this.source],
									folder:this.activeFolder,
									contents:this.nodes,
									view:this.activeView
								};

								if(tapLength>0&&tapLength<300){
									this.lastTap=0;
									return void this.defaultFileAction.exec(actionContext)
								}

								this.lastTap=currentTime;
								Us?.enabled?.(actionContext)&&Us.exec(actionContext)
								/* kmish  ends */
								
                            },
```
